--- controllers/footbot_diffusion/footbot_diffusion_original.cpp	2026-06-27 22:22:15.164443000 -0400
+++ controllers/footbot_diffusion/footbot_diffusion.cpp	2026-06-28 15:58:18.890718910 -0400
@@ -62,34 +62,25 @@
 /****************************************/
 
 void CFootBotDiffusion::ControlStep() {
-   /* Get readings from proximity sensor */
-   const CCI_FootBotProximitySensor::TReadings& tProxReads = m_pcProximity->GetReadings();
-   /* Sum them together */
-   CVector2 cAccumulator;
-   for(size_t i = 0; i < tProxReads.size(); ++i) {
-      cAccumulator += CVector2(tProxReads[i].Value, tProxReads[i].Angle);
-   }
-   cAccumulator /= tProxReads.size();
-   /* If the angle of the vector is small enough and the closest obstacle
-    * is far enough, continue going straight, otherwise curve a little
-    */
-   CRadians cAngle = cAccumulator.Angle();
-   if(m_cGoStraightAngleRange.WithinMinBoundIncludedMaxBoundIncluded(cAngle) &&
-      cAccumulator.Length() < m_fDelta ) {
-      /* Go straight */
-      m_pcWheels->SetLinearVelocity(m_fWheelVelocity, m_fWheelVelocity);
-   }
-   else {
-      /* Turn, depending on the sign of the angle */
-      if(cAngle.GetValue() > 0.0f) {
-         m_pcWheels->SetLinearVelocity(m_fWheelVelocity, 0.0f);
-      }
-      else {
-         m_pcWheels->SetLinearVelocity(0.0f, m_fWheelVelocity);
-      }
+   /*fFetch sensor readings and initialize thearepulsion vector */
+   const CCI_FootBotProximitySensor::TReadings& tProx = m_pcProximity->GetReadings();
+   CVector2 cRepulsionVec;
+
+   /* Compute vector sum inspired in repulsive forces based on the potential field method */
+   for(size_t i = 0; i < tProx.size(); ++i) {
+      cRepulsionVec += CVector2(tProx[i].Value, tProx[i].Angle + CRadians::PI);
    }
-}
 
+   /* Define the heading direction by combining forward motion with potential field correction */
+   CVector2 cSteering = CVector2(1.0f, CRadians::ZERO) + (cRepulsionVec * 2.0f);
+   
+   /* Calculate wheel velocities based on the angular steering error */
+   Real fTurn = cSteering.Angle().GetValue();
+   
+   /* Map  error to wheel velocities to maintain continuous motion(simple-elegant method) */
+   m_pcWheels->SetLinearVelocity(m_fWheelVelocity * (1.0f - std::abs(fTurn)), 
+                                 m_fWheelVelocity * (1.0f + fTurn));
+}
 /****************************************/
 /****************************************/
 
